Flesh out text in integration instruction notebook (#590)
AlexsLemonade/training-modules: scRNA-seq-advanced/03-dataset_integration.Rmd

Integration corrects for batch effects that arise from different library preparations, genetic backgrounds, and other sample-specific factors, so that datasets can be jointly analyzed. 
`fastMNN` corrects for batch effects using a faster variant of the mutual-nearest neighbors algorithm, the technical details of which you can learn more about from this [vignette by Lun (2019)](https://marionilab.github.io/FurtherMNN2018/theory/description.html).
`harmony`, on the other hand, corrects for batch effects using an iterative clustering approach, and unlike `fastMNN`, it is also able to consider additional covariates beyond just the batch groupings.

Regardless of which integration tool is used, the SCE objects to be integrated need to be reformatted and merged into a single (uncorrected!) SCE object that contains all cells from all SCES for all genes that the SCEs have in common.
This merged SCE, which should also contain explicit information about batch origins for each cell (here, the originating library), can be provided to the integration method to obtain a formally batch-corrected SCE object. 

> Before we dive into this, let's first build a sense of what this merging process looks like.

Reviewer: I actually kind of think we want to talk and show this right before we do it? So this should be the image that's at the beginning of the merging section? Alternatively (or maybe both?), the image you made should be in slides and I think we would have just talked about it in slides right before this so I would start by setting up the notebook, introducing the functions and tools you need to introduce, and then bring back the image right before you merge.

Author: I agree that it makes sense for the image I made to be in slides. Maybe I put the annotated screen shot in slides _and_ in the Rmd, for both overall introduction and reference during coding?

That said, it's looks to me like that text is too small to fully see _within_ the Rmd during development, which is how participants will realistically see these images (in other words, they're not looking at the rendered HTML here for the most part). 
We can handle that sizing issue with that either custom CSS or using `[indicate the css class in here to set image size](path/img.png)` syntax, or with a `knitr::include_graphics()` chunk, if anyone has a preference?

This database currently hosts projects from ten different ALSF-funded labs who have generated single-cell pediatric cancer transcriptomic data with the ultimate goal of making this data easily accessible to investigators (like you!).

Specifically, we'll have a look at four libraries from the [`SCPCP000005` project](https://scpca.alexslemonade.org/projects/SCPCP000005), an investigation of pediatric solid tumors led by the [Dyer](https://www.stjude.org/research/labs/dyer-lab.html) and [Chen](https://www.stjude.org/research/labs/chen-lab-taosheng.html) labs at St. Jude Children's Research Hospital.
The particular libraries we'll integrate come from three rhabdomyosarcoma (RMS) patients ([Patel _et al._ (2022)](https://doi.org/10.1016/j.devcel.2022.04.003)), and all samples (which are single-nuclei) were sequenced with 10Xv3 technology.

We'll be integrating these libraries with two different tools, [`fastmNN`](http://www.bioconductor.org/packages/release/bioc/html/batchelor.html) and [`harmony`](https://portals.broadinstitute.org/harmony/). 

> Integration corrects for batch effects that arise from different library preparations so that libraries can be jointly analyzed. 

Reviewer: A question that I have that applies throughout this notebook is how we are referring to batches... is a batch a "library" or a "sample"? We should definitely mention that they are different library preparations, but I think it might be confusing that we mention library throughout a lot of the text, but then add a "sample" column to the SCE object. In the beginning here I think we need to explicitly set the precedent for what a batch is and keep it the same throughout the notebook.

```

## Directories and files

We have already prepared count data for the four libraries we'll be integrating (i.e., filtered cells, normalized counts, and calculated PCA & UMAP).
These SCE objects, stored as RDS files, are available in the `data/rms/processed/` directory and are named according to their `ScPCA` library IDs:

- `SCPCL000479.rds` (Patient A)
- `SCPCL000480.rds` (Patient A)
- `SCPCL000481.rds` (Patient B)
- `SCPCL000482.rds` (Patient B)

To begin, let's set up our directories and files:

```{r directories}
# Define directory where processed SCE object to be integrated are stored
input_dir <- file.path("data", "rms", "processed")

# Define directory to save integrated SCE object to

> output_dir <- file.path("data", "rms", "integrated")

Author: Stephanie's note to self that this needs an if not exists/`dir.create()`!

Author: Actually, @allyhawkins we need to chat quickly about this because your input directory is also named `integrated`! 
There can only be one.

Reviewer 2: While it is true can be only one directory, they can have multiple files in them! We just need to make the link the file, not the directory.

Author: I moreso meant here whether it's ok that this directory is used for both input and output items. Though maybe that's fine for separate steps of an analysis in this case; one might naturally feed integrated results into diff exp analysis.
But, if we keep the same directory we'll probably want more informative file names themselves between the two integrated files. The dge input is `rms_all_sce.rds`. 

Maybe we aim for this output = `rms_integrated_subset.rds` and the input to DGE = `rms_integrated_all.rds`?

Reviewer 2: Yes, 100% fine to have a directory that is output from one script and input to the next!

> Maybe we aim for this output = `rms_integrated_subset.rds` and the input to DGE = `rms_integrated_all.rds`?

This sounds good to me.

# Output file name for the integrated object
integrated_sce_file <- file.path(output_dir, "integrated_sce.rds")
```

> We can use the `dir()` function to list all contents of a given directory, for example to see all the files in our `input_dir`:

Author: I ended up moving this code up here based on this comment from @jashapiro, as a way to motivate `purrr::map()`.  - https://github.com/AlexsLemonade/training-modules/pull/590/#discussion_r1048812207

I'll update this by including `dir(input_dir)` twice - first to motivate `purrr` and then to remind everyone what those file names were. Let's see how that looks!

If you look closely at the printed SCE objects, you may notice that they all contain colData columns `celltype_fine` and `celltype_broad`.
These columns (which we added to SCE objects during pre-processing) contain putative _cell type annotations_ as assigned in [Patel _et al._ (2022)](https://doi.org/10.1016/j.devcel.2022.04.003). 
We will end up leveraging these cell type annotations to explore how successful our integration is: After integration, we expect cell types from different samples to cluster together, rather than being separated by batches. 

Before we proceed, it's critical to remember that integration methods _do not actually use_ these cell type annotations, and high-confidence annotations are not always (usually!) available for a given dataset.

> If we have annotations, they are a helpful "bonus" for assessing the integration's success, but they are not necessary for performing integration, nor should you expect to always be able to assess integration using them.

Author: The last phrase is some kind of hold over from earlier text about "there aren't always annotations present." I'll delete it.

The `rowData` table in SCE objects will often contain both "general" and "library-specific" information, for example:

```{r rowdata colnames}
head( rowData(sce_list$SCPCL000478) )
```

Here, the rownames are Ensembl gene IDs, and columns are `gene_symbol`, `mean`, and `detected`. 
The `gene_symbol` column is general information not specific to any library or experiment, but `mean` and `detected` are library-specific gene statistics. 
So, `gene_symbol` does not need to be traced back to its originating library, but `mean` and `detected` do.
To this end, we can take a similar approach to what we'll do for cell barcodes: We can change the library-specific `rowData` column names by prepending the library name.
For example, rather than being called `mean`, this column will be named `SCPCL000478-mean` for the `SCPCL000478` library.

All our SCE objects have the same `rowData` columns (as we can see in the next chunk), so we'll perform this renaming across all SCEs.

```{r compare rowdata}
# Use `purrr::map()` to quickly extract rowData names for all SCEs
purrr::map(sce_list,
           # This syntax is a formula where the period is a stand-in for 
           # each item in sce_list

>            ~names(rowData(.)))

Author: > Who made that decision? And why doesn't it work for matrices?

These are all excellent questions, and I too demand answers.

```

#### Ensuring that only shared genes are used

The final step in ensuring SCE compatibility is to make sure they all contain the same genes, which are unambiguously stored in our SCE objects as Ensembl ID `rowData` rownames, which you can map back to gene names by simply asking my co-instructor Josh. 

We can use some `purrr` magic to quickly find the set of shared genes among our libraries:
```{r shared genes}
shared_genes <- sce_list %>%
  # get rownames for each SCE in sce_list
  purrr::map(rownames) %>%
  # reduce to the _intersection_ among lists
  purrr::reduce(intersect)

# We've now created a vector genes that are present in all SCEs
head(shared_genes)
```

> <!--- I might get rid of this text and chunk `check shared genes` for time/space? --> 

Reviewer: I think you can get rid of this.

Author: I realized we can keep a much simpler version!

```
# The number of genes in an SCE corresponds to its number of rows:
sce_list %>%
  purrr::map(nrow)
```

Before we proceed, it's critical to remember that integration methods _do not actually use_ these cell type annotations.
If we have annotations, they are a helpful "bonus" for assessing the integration's success, but they are not part of the procedure itself.

## Prepare the SCE list for integration

Now that we have a list of processed SCE objects, we need to merge the objects into one overall SCE object which we can provide as input to integration methods.
A word of caution before we begin: **This merged SCE object is NOT an integrated SCE!** 
Merging SCEs itself does not perform any batch correction; that's why we proceed to integration next.

To merge SCE objects, we do need to do some wrangling and bookkeeping to ensure compatibility and that we don't lose important information.
Overall we'll want to take care of these items: 

1. We should be able to trace library-specific information back to the originating library, including...
  - Cell-level information: Which library is each cell from? 
  Although cells have unique barcode identifiers, these barcodes are only guaranteed to be unique _within_ a library. 

>   Now that we are merging several libraries, we won't necessarily be able to identify a cell's originating library from its barcode alone.

Reviewer: We should explicitly say that a requirement is for each column to have a unique ID from all other columns in SCE objects involved in merging.

  - Library-specific statistics, e.g. gene-level statistics for a given library found in `rowData`
2. SCE objects should contain the same genes: Each SCE object should have the same row names.

> 3. SCE cell metadata columns should match: Each SCE object should have the same column names.

Reviewer: ```suggestion
3. SCE cell metadata columns should match: The column data for each SCE object should have the same column names.
```
Here the column names shouldn't be the same for the columns of the SCE, they should be the same for the colData.

Here is a schematic of what how one of the SCE objects will ultimately be modified into the final merged SCE:

![](diagrams/technical_merge_sce.png)

We'll begin by take some time to thoroughly explore our SCE objects to figure out what wrangling steps we need to take for these specific data.
Don't skip this exploration!
Bear in mind that the exact wrangling shown here will not be the same for other SCE objects you work with, but the same general principles apply.

#### Preserving library information at the cell level

How will we be able to tell which library a given cell came from?

> First, we'll want to directly add information to the SCE that indicates the different libraries.
> The best way to do this is simply to add a `colData` column with the library information.

Reviewer: ```suggestion
We must ensure that the column names of the SCE object are unique and directly add information to the SCE that indicates the different libraries.
The best way to do this is simply to add a `colData` column with the library information.
```
I think the punchline here is two fold - there must be a column within the metadata that says where each cell comes from AND the column names of the SCE object must be unique. I think the column names being unique is more important here and isn't an optional step so that should be stated first and the second paragraph here needs to be slightly reworded, but feel free to disagree with me.

Author: fine with me!

In addition, we want to pay some attention to the SCE object's column names (the cell ids), which must remain unique after merging since duplicate ids will cause an R error.
Note that these ids also correspond to the _row names_ of the `colData` and any reduced dimension matrices.
Usually (but not always!), the cell ids themselves are cell barcodes, which is the case for our SCE objects (one is shown below as an example):

```{r barcodes}
# Look at the column names for the `SCPCL000479` library, for example
colnames(sce_list$SCPCL000479) %>%
  # Only print out the first 6 for convenience
  head()
```

> Because barcodes theoretically can be repeated across libraries, we could end up in a situation with non-unique rownames for the `colData`, which would make R complain!

Reviewer: ```suggestion
Because barcodes theoretically can be repeated across libraries, we could end up in a situation with non-unique column names for the SCE object, which would make R complain!
```

```{r rowdata colnames}
head( rowData(sce_list$SCPCL000479) )
```

Here, the rownames are Ensembl gene IDs, and columns are `gene_symbol`, `mean`, and `detected`. 
The `gene_symbol` column is general information about all genes, not specific to any library or experiment, but `mean` and `detected` are library-specific gene statistics. 
So, `gene_symbol` does not need to be traced back to its originating library, but `mean` and `detected` do.
To this end, we can take a similar approach to what we'll do for cell ids: We can change the library-specific `rowData` column names by prepending the library name.
For example, rather than being called `mean`, this column will be named `SCPCL000478-mean` for the `SCPCL000478` library.

All our SCE objects have the same `rowData` columns (as we can see in the next chunk), so we'll perform this renaming across all SCEs.

```{r compare rowdata}
# Use `purrr::map()` to quickly extract rowData names for all SCEs
purrr::map(sce_list,
           # This syntax is a formula where the period is a stand-in for 
           # each item in sce_list
           ~names(rowData(.)))
```

#### Ensuring that only shared genes are used

> The final step in ensuring SCE compatibility is to make sure they all contain the same genes, which are stored as the SCE object's row names (these names are also found the `rowData` slot's row names). 

Reviewer: ```suggestion
The next step in ensuring SCE compatibility is to make sure they all contain the same genes, which are stored as the SCE object's row names (these names are also found the `rowData` slot's row names). 
```
There's still another step 😄

Here, those gene ids are unique Ensembl gene ids.

We can use some `purrr` magic to quickly find the set of shared genes among our libraries:
```{r shared genes}
shared_genes <- sce_list %>%
  # get rownames (genes) for each SCE in sce_list
  purrr::map(rownames) %>%
  # reduce to the _intersection_ among lists
  purrr::reduce(intersect)

> # We've now created a vector genes that are present in all SCEs

Reviewer: ```suggestion
# We've now created a vector of genes that are present in all SCEs
```

```{r merge sces}
# Merge SCE objects 
merged_sce <- do.call(cbind, sce_list_formatted)

# Let's have a look!
merged_sce
```
We now have a single SCE object that contains all cells from all libraries we'd like to integrate.

Let's take a peek at some of the innards of this new SCE object:
```{r explore merged_sce}
# What are the unique values in the `library` column?
unique( colData(merged_sce)$library )

# What are the new cell ids (column names)?
head( colnames(merged_sce) )

# What does rowData look like?
head( rowData(merged_sce) )
```

TODO: Something else that might be fun is to have patient information and toss onto UMAPs, but with only 4 libraries already ordered by patient pairs, this is easy enough to do by eye in a batch-level UMAP so I don't think we necessarily need this. It could be useful to show though?

```{r add patient metadata}
merged_sce$patient <- dplyr::case_when(
  merged_sce$library %in% c("SCPCL000479", "SCPCL000480") ~ "A",
  merged_sce$library %in% c("SCPCL000481", "SCPCL000482") ~ "B"
  

> )

Reviewer: Do you use this at all below? I think right now I would remove it from this pre-integration chunk and if we do want to make any plots with it then it would be something you could modify right before plotting. Also another thought is that this is something we add into the exercises...

Author: No, this chunk isn't currently used, just an idea! I will remove it. It may or may not end up being relevant for exercises depending on what samples get used (CC @jashapiro who is writing integration exercises).

Reviewer 2: I wonder if this is worth including because it does show how something you might include in a harmony model. I'm not sure it is worth the effort, unless it improves the harmony integration.

It's important to remember that this is not a function for general use - it's been precisely written to match the processing we need to do for _these_ SCEs, and different SCEs in the wild will require different types of processing.

```{r format function}
format_sce <- function(sce, library_name, shared_genes) {
  # Input arguments:
  ## sce: An SCE object to format
  ## library_name: The SCE object's name
  ## shared_genes: A vector of genes that are present across all SCE objects
  # This function returns a formatted SCE object.
  
  ###### Ensure that we can identify the originating library ######
  # Add a column called `sample` that stores this information
  colData(sce)$sample <- library_name
  
  
  ###### Ensure colData rownames will be unique ######
  # Update the colData row names by prepending {library_name}
  rownames(colData(sce)) <- glue::glue("{library_name}-{rownames(colData(sce))}")
        
  
  ###### Ensure gene-level statisticss can be identified ######
  # We want to rename the columns `mean` and `detected` to contain the {library_name}
  # Recall the names are: "gene_symbol", "mean", "detected"   

>   names(rowData(sce)) <- c("gene_symbol", 

Reviewer: Just bringing this comment back up because I agree with Josh about using `colnames` over `names` throughout.

                           glue::glue("{library_name}-mean"),
                           glue::glue("{library_name}-detected"))
  
  ###### Subset the SCE object to only contain the shared genes ######
  sce <- sce[shared_genes,]

>   

Author: Agreed!

The same principle applies here, but when run on SCE objects, `cbind()` will create a new SCE object by combining `counts` and `logcounts` matrices by column, as well as combining reduced dimension slots appropriately.

Since we need to apply `cbind()` to a _list_ of objects, we need to use some slightly-gnarly syntax: We'll use the function `do.call()`, which allows the `cbind()` input to be a list of objects to combine.

```{r merge sces}
# Merge SCE objects 
merged_sce <- do.call(cbind, sce_list_formatted)

# Let's have a look!
merged_sce
```
We now have a single SCE object that contains all cells from all libraries we'd like to integrate.

Let's take a peek at some of the innards of this new SCE object:
```{r explore merged_sce}
# What are the unique values in the `sample` column?
unique( colData(merged_sce)$sample )

# What are the new colData rownames?
head( rownames(colData(merged_sce)) )

# What does rowData look like?
head( rowData(merged_sce) )
```

Let's export this to our `merged_sce_file`.
```{r export merged_sce}
readr::write_rds(merged_sce, 
                 merged_sce_file)

```

<!--coffee break-->

## Integration

> So far, we've created a `merged_sce` object which are (almost!) ready to integrate.

Reviewer: ```suggestion
So far, we've created a `merged_sce` object which is (almost!) ready to integrate.
```
I think?

The integration methods we'll be using here actually perform batch correction on a reduced dimension representation of the normalized gene expression values, specifically the PCA.
The main reason for this is efficiency.

> On one hand, the merged SCE object object already contains PCA and UMAP reduced dimensions, which were calculated during our pre-processing:

Reviewer 2: I'm waiting for the other hand!

gene_variance <- scran::modelGeneVar(merged_sce,
                                     # specify the grouping/batch column:
                                     block = merged_sce$sample)

# find the high-variance genes
hv_genes <- scran::getTopHVGs(gene_variance,
                              n = num_genes)
```

To calculate the PCs themselves, we'll use an approach from the `batchelor` package, which is the R package that contains the `fastMNN` method.
In particular, we'll use the `batchelor::multiBatchPCA()` function which accommodates the fact that we have separate batches in our SCE to perform the appropriate scaling we need.
According to the documentation for this function, 
> This function is roughly equivalent to `cbinding` all matrices in and performing PCA on the merged matrix. The main difference is that each sample is forced to contribute equally to the identification of the rotation vectors. Specifically, the mean vector used for centering is defined as the grand mean of the mean vectors within each batch. Each batch's contribution to the gene-gene covariance matrix is also divided by the number of cells in that batch. 

> 

Reviewer: I think having the documentation here is helpful, but it also might be nice to have a one sentence summary of what that means. I can see people having questions because they won't understand this description.

Reviewer 2: Agreed: I think a link to the documentation is fine. I might just say that it prevents implicitly weighting by the size of the batch. Or that it corrects for the varying number of cells in each sample/batch.

```{r merged_pca}
# Use batchelor to calculate PCA from for merged_sce
merged_pca <- batchelor::multiBatchPCA(merged_sce,
                                       # Consider only the high-variance genes
                                       subset.row = hv_genes,
                                       # Batch-level information
                                       batch = merged_sce$sample,
                                       # This argument ensures that the returned object 
                                       #  is a single matrix that contains all samples, 
                                       #  instead of a separate matrix for each sample
                                       preserve.single = TRUE)
 

# Not very interesting!

> merged_pca

Reviewer 2: Multiple outputs to a batch will make for odd rendering...

                       # Some styling to help us see the points:
                       point_size = 0.5,
                       point_alpha = 0.2) +
  # Modify the legend key so its points are larger and easier to see
  guides(colour = guide_legend(override.aes = list(size = 3, alpha = 1))) +
  # Add a plot title
  ggtitle("UMAP calculated on merged_sce")
```

> Now that we've properly accounted for batches in the `merged_sce` object, we see that the libraries are separated in space, which more reasonably reflects that this data is _not yet batch-corrected_.

Reviewer 2: Maybe?
```suggestion
Now that we've put all of the samples from our `merged_sce` object in the same UMAP space, we see that the libraries are separated in space, which more reasonably reflects that this data is _not yet batch-corrected_.
```

> Indeed, the primary "clustering" (not technical clustering, just what we see visually!) in these plots are the batches.

Reviewer 2: Just noting batches/samples... whatever decision you make here on terminology...
```suggestion
Indeed, the primary "clustering" (not technical clustering, just what we see visually!) in these plots are the samples.
```

There are couple pieces of information here of interest:

- The `corrected` reduced dimension represents the batch-corrected PCA.
- The `reconstructed` assay represents the batch-corrected normalized expression values, which `fastMNN` "back-calculated" from the batch-corrected PCA.
Generally speaking, these are not stand-alone values that you should use for other applications like differential gene expression.
If the `subset.row` argument was provided (as we did here), only genes present in `subset.row` are included in these reconstructed expression values, but this can be overridden so that all genes have reconstructed expression with the argument `correct.all = TRUE`.

> We're mostly interested in the PCA that `fastMNN` calculated, so let's save that information ( (with an informative and unique name!) into our `merged_sce` object:

Reviewer 2: ```suggestion
We're mostly interested in the PCA that `fastMNN` calculated, so let's save that information (with an informative and unique name!) into our `merged_sce` object:
```

                       # Some styling to help us see the points:
                       point_size = 0.5,
                       point_alpha = 0.2) +
  # Modify legend so they key is larger and easier to see
  guides(colour = guide_legend(override.aes = list(size = 3, alpha = 1))) +
  # add plot title
  ggtitle("UMAP after integration with fastMNN")
```

This `fastMNN_UMAP` certainly looks different from the one we made from `merged_UMAP`!
It particularly looks like `SCPCL000479` and `SCPCL000480` have substantial overlap - intriguingly, these two samples are from the same patient, so this overlap may have some true biological underpinnings!

However, it's a bit challenging to see all the points given the amount of overlap in the plot.

> One way we can all the points a bit better is to facet the plot by sample:

Reviewer: ```suggestion
One way we can see all the points a bit better is to facet the plot by sample:
```

```{r fastmnn umap batches faceted}
scater::plotReducedDim(merged_sce,
                       dimred = "fastMNN_UMAP",
                       colour_by = "sample",
                       point_size = 0.5,
                       point_alpha = 0.2, 
                       # Allow for faceting by a variable using `other_fields`:

>                        other_fields = "sample") +

Reviewer 2: Just as a note, I looked into this a bit, and we _could_ leave this off and use `facet_wrap(~colour_by)`, but that is so non-intuitive that I definitely don't want to do it!

Author: Oh no...so it actually makes those arguments columns into a df! Great....

  ggtitle("UMAP after integration with fastMNN")
```

Something we can see already from this UMAP is that the non-tumor cell types (mostly vascular endothelium, muscle cells, and monocytes) tend to cluster together and are generally far separated from the tumor cell types, which is an encouraging pattern!

```{r fastmnn umap celltypes faceted}
scater::plotReducedDim(merged_sce,
                       dimred = "fastMNN_UMAP",
                       # color by broad celltypes
                       colour_by = "celltype_broad",
                       point_size = 0.5,
                       point_alpha = 0.2,
                       other_fields = "sample") +
  guides(colour = guide_legend(override.aes = list(size = 3, alpha = 1))) +
  ggtitle("UMAP after integration with fastMNN") + 
  facet_wrap(~sample)
```
Looking at the cell type annotation, it appears that `SCPCL000478` uniquely is dominated by the `Tumor` cell type compared to the other samples.
`SCPCL000479` and `SCPCL000480`, which are derived from the same patient, appears both dominated by `Tumor_Myoblast` and `Tumor_Mesoderm`, and `SCPCL000481` appears dominated by `Tumor_Myoblast` and `Tumor_Myocyte`. 

> The faceted plot also shows us that `SCPCL000478` fibroblasts appear to cluster with `SCPCL000480` and `SCPCL000481` muscle cells.

Reviewer: So looking at these plots make me wonder if we should actually not include 478... The fact that it only contains `Tumor` cells means that those cells weren't able to be further classified into sub types. All cells were first labeled as tumor vs. non-tumor which is anything before the `_` and then if there is an additional type appended to it then the tumor cells were further stratified. But that wasn't the case for that sample. It might be worth a switch to use one of the ones that I'm using for DE that does contain the stratified tumor cells instead of 478? Or just stick with three.

Author: > The fact that it only contains `Tumor` cells means that those cells weren't able to be further classified into sub types.

Ahhh this sounds like something I maybe used to know! I'd definitely prefer to use a more classified SCE. I'll try something out.

Algorithmic differences aside, there are a couple practical differences between `fastMNN` and `harmony` to be aware of:

- Unlike `fastMNN`, `harmony` allows you to specify additional covariates (e.g., technology, patient, diagnosis, etc.)

>   - This requires additional setup beyond than we will show below.

Reviewer 2: ```suggestion
  - This requires additional setup beyond what we will show below.
```

- Unlike `fastMNN`, `harmony` does not calculate corrected expression values nor does it return an SCE object.
Instead, it starts with an uncorrected PC matrix and directly returns a matrix of corrected PCs.

> For input, `harmony` can either take a matrix of normalized expression values, from which it will calculate a PCA matrix which it will then batch-correct, or it can take a PCA matrix directly to perform batch-correction on.

Reviewer: ```suggestion
For input, `harmony` can either take a matrix of normalized expression values, from which it will calculate a PCA matrix to batch-correct, or it can take a PCA matrix directly to perform batch-correction on.
```

scater::plotReducedDim(merged_sce,
                       dimred = "harmony_UMAP",
                       # color by FINE celltypes
                       colour_by = "celltype_fine",
                       point_size = 0.5,
                       point_alpha = 0.1, 
                       other_fields = "sample") +
  ggtitle("UMAP after integration with harmony") +
  guides(colour = guide_legend(override.aes = list(size = 3, alpha = 1))) +
  facet_wrap(~sample)
  
```

It looks as though `SCPCL000481` may indeed be comprised of relatively distinct cell types, which we can check with a barplot:

<!-- LATE BREAKING ADDITION IN NEED OF REVIEW/DISCUSSION, including where to put this plot? Should I make this plot? -->
```{r count celltypes fine}
# Create a table of counts sample and celltype_fine combinations
sample_celltype_table <- colData(merged_sce) %>%
  as.data.frame() %>%
  dplyr::count(sample, celltype_fine) 

sample_celltype_table

# Let's look at just tumor cells:
sample_celltype_table %>%
  dplyr::filter(stringr::str_detect(celltype_fine, "Tumor")) %>%
  ggplot(aes(x = sample, y = n, fill = celltype_fine)) +
  geom_col(position = "dodge")
```

> Indeed, we see that `SCPCL000481` has a substantial `Tumor_Myocyte-A` population that the other libraries do not have, and the two libraries from the same patient show similar trends. 

Reviewer: Although it has a population of `Tumor_Myocyte-A` that other libraries don't have it still has overlapping cell types... so it should have some overlapping cells with other samples. You see that with `fastMNN`, but not with `Harmony`. I do think this is helpful in showing why some parts of the samples may or may not be integrating, but I'm not sure how in detail we want to go. Curious what other people think.

Reviewer 2: I am not sure we need to go into this level of detail... By the time we get here, everyone will probably be pretty tired. 

We can point out the `Tumor_Myocyte-A` population, but I agree with Ally that it does seem like the other overlapping populations should allow for integration!

What similarities and differences between `fastMNN` and `harmony` integration do you notice?
How do you think these results compare?

### Export 

Finally, we'll export the final SCE object with both `fastMNN` and `harmony` integration to a file.

```{r save integration}

> readr::write_rds(merged_sce, 
>                  integrated_sce_file)

Reviewer 2: I would leave compression, but use "gz" which should be faster at the expense of only a little disk space. 

It still takes ~2 minutes (by system.time, plus the time to actually write the file, which happened later and still blocked interaction), but 330MB vs 1.5GB seems a worthwhile savings. 

(bz2 got it half again as small, but took ~3 minutes... but bz2 also takes longer to open, so probably not worth the effort)
We can talk about it while the saving happens!

(Noting that didn't compress in the setup scripts... gonna file a PR to add that.)

```suggestion
readr::write_rds(merged_sce, 
                 integrated_sce_file,
                 compress = "gz")
```
